[#10] 언어 추가 기능 구현

msugo1 opened the pull request:
## Model 관련

### LanguageData 추가
- 언어 등록 요청으로 들어오는 JSON 데이터를 매핑
- 언어 이름`name` 과 레벨`level`을 멤버값으로 가지고 있음

## Enum 클래스 관련

### LanguageLevel 추가
- `언어 레벨`에 대한 상수값, 카테고리 값을 가지고 있는 Enum 클래스
- DB에 저장시 카테고리 int 값으로 변경해서 저장하기 위해 TypeHandler를
  static nested class로 가지고 있음

### LanguageStatus 추가
- `언어 학습 상태`를 표현할 상수값, 카테고리 값을 가지고 있는 Enum 클래스
- LangaugeLevel과 같은 이유로 TypeHandler를 static nested class로 가지고 있음

### Language 추가
- 각 나라의 `언어 이름들`을 상수값으로 담고 있는 Enum 클래스

## Controller 계층 관련

### LanguageController 추가
- 언어등록에 대한 요청을 받을 `addLanguages` 메소드 추가

## Service 계층 관련

### LanguageService 추가
- 언어 추가 로직을 다룰 `addLanguages` 메소드 추가
1. db 내에서 추가하려는 언어의 `status` 로 등록된 언어 개수를 가져온 뒤
   이를 요청 받은 언어의 수와 비교해서 언어 추가 제한을 넘지 않는지 확인
2. status 에 맞게 레벨이 잘 설정 되었는지 확인
(모국어는 모국어 레벨만 가질 수 있으며, 모국어가 아닌 경우는 모국어 레벨을 설정할 수
없음)
3. 기존에 추가된 언어와 새로 요청받은 언어 간의 중복 추가 비교

## Repository/Mapper 계층 관련

### LanguageRepository/LanguageMapper 추가
- 요청 받은 언어들 DB에 추가하기 위한 `insertLanguages` 메소드 추가
- 해당 언어 학습 status 로 저장된 총 언어수를 가져오기 위한
  `countLanguages` 메소드 추가
- 현재 로그인 중인 사용자의 id를 가지고 해당 사용자의 언어 목록을
  가져오기 위한 `getLanguages` 메소드 추가

### LanguageMapper.xml 추가
- 위에서 추가한 메소드에 대한 쿼리 내용을 담고 있는 mapper.xml 파일

## Exception 관련

### DuplicateLanguageException 추가
- 한 언어에 대한 `중복추가` 요청이 들어올 경우 처리하기 위한 예외 클래스

### InvalidLanguageException 추가
- `적절하지 않은 레벨`을 가진 언어 추가 요청을 처리하기 위한 예외 클래스

### LanguageLimitExceededException 추가
- 해당 언어 학습 상태 별로 `최대 추가 언어 수의 제한`이 있으며 이를
  `초과했을 때` 발생할 예외를 처리하기 위한 예외 클래스

### 해당 예외들을 ControllerExceptionHandler에 매핑
- 위의 예외가 발생 시 예외 메시지와 함께 `Bad Request(400)`을 사용자에게 리턴

## Handler 관련

### EnumCategoryTypeHandler 추가
- DB에 Enum 상수 값을 저장시 상수가 담고 있는 category 정수 값으로 변환해
  저장하고 다시 DB에서 값을 받아올 때는 Enum 상수 값으로 자동 변환해서
  가지고 오는 `TypeHandler` 추가

## 테스트 관련

### Service 계층 `LanguageService` 테스트 작성
- LanguageService의 addLangauges 메소드에 대한 예상 가능한 예외/성공 케이스에 대한
  단위, 통합테스트 작성 및 구동
- Postman 을 활용해 예상 가능한 예외/성공 케이스에 대해 해당 로직이 제대로 작동하는지 확인

Commit:
[#10] 언어 추가 기능 구현

Model 관련

* LanguageData 추가
- 언어 등록 요청으로 들어오는 JSON 데이터를 매핑
- 언어 이름`name` 과 레벨`level`을 멤버값으로 가지고 있음

Enum 클래스 관련

* LanguageLevel 추가
- `언어 레벨`에 대한 상수값, 카테고리 값을 가지고 있는 Enum 클래스
- DB에 저장시 카테고리 int 값으로 변경해서 저장하기 위해 TypeHandler를
  static nested class로 가지고 있음

* LanguageStatus 추가
- `언어 학습 상태`를 표현할 상수값, 카테고리 값을 가지고 있는 Enum 클래스
- LangaugeLevel과 같은 이유로 TypeHandler를 static nested class로 가지고 있음

* Language 추가
- 각 나라의 `언어 이름들`을 상수값으로 담고 있는 Enum 클래스

Controller 계층 관련

* LanguageController 추가
- 언어등록에 대한 요청을 받을 `addLanguages` 메소드 추가

Service 계층 관련

* LanguageService 추가
- 언어 추가 로직을 다룰 `addLanguages` 메소드 추가
1. db 내에서 추가하려는 언어의 `status` 로 등록된 언어 개수를 가져온 뒤
   이를 요청 받은 언어의 수와 비교해서 언어 추가 제한을 넘지 않는지 확인
2. status 에 맞게 레벨이 잘 설정 되었는지 확인
(모국어는 모국어 레벨만 가질 수 있으며, 모국어가 아닌 경우는 모국어 레벨을 설정할 수
없음)
3. 기존에 추가된 언어와 새로 요청받은 언어 간의 중복 추가 비교

Repository/Mapper 계층 관련

* LanguageRepository/LanguageMapper 추가
- 요청 받은 언어들 DB에 추가하기 위한 `insertLanguages` 메소드 추가
- 해당 언어 학습 status 로 저장된 총 언어수를 가져오기 위한
  `countLanguages` 메소드 추가
- 현재 로그인 중인 사용자의 id를 가지고 해당 사용자의 언어 목록을
  가져오기 위한 `getLanguages` 메소드 추가

* LanguageMapper.xml 추가
- 위에서 추가한 메소드에 대한 쿼리 내용을 담고 있는 mapper.xml 파일

Exception 관련

* DuplicateLanguageException 추가
- 한 언어에 대한 `중복추가` 요청이 들어올 경우 처리하기 위한 예외 클래스

* InvalidLanguageException 추가
- `적절하지 않은 레벨`을 가진 언어 추가 요청을 처리하기 위한 예외 클래스

* LanguageLimitExceededException 추가
- 해당 언어 학습 상태 별로 `최대 추가 언어 수의 제한`이 있으며 이를
  `초과했을 때` 발생할 예외를 처리하기 위한 예외 클래스

* 해당 예외들을 ControllerExceptionHandler에 매핑
- 위의 예외가 발생 시 예외 메시지와 함께 `Bad Request(400)`을 사용자에게 리턴

Handler 관련

* EnumCategoryTypeHandler 추가
- DB에 Enum 상수 값을 저장시 상수가 담고 있는 category 정수 값으로 변환해
  저장하고 다시 DB에서 값을 받아올 때는 Enum 상수 값으로 자동 변환해서
  가지고 오는 `TypeHandler` 추가

테스트 관련

* Service 계층 `LanguageService` 테스트 작성
- LanguageService의 addLangauges 메소드에 대한 예상 가능한 예외/성공 케이스에 대한
  단위, 통합테스트 작성 및 구동
- Postman 을 활용해 예상 가능한 예외/성공 케이스에 대해 해당 로직이 제대로 작동하는지 확인

Commit:
[#10] 주석 중복내용 수정

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/LanguageController.java:
```diff
@@ -0,0 +1,35 @@
+
+        languageService.addLanguages(userId, newLangData, status);
+
+        return HTTP_RESPONSE_OK;
```
이것도 생략이 가능해보이네요~

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/LanguageController.java:
```diff
@@ -0,0 +1,35 @@
+    @PostMapping
+    public ResponseEntity<Void> addLanguages(@CurrentUser String userId,
+                                            @RequestBody List<LanguageData> newLangData,
+                                            @RequestParam LanguageStatus status) {
```
이거는 `@RequestParam`으로 따로 나눠주신 이유가 있으실까요?

msugo1 replied:
포스트맨에서 데이터를 `POST`로 보낼 때 문제가 있어서 `@RequestParam`으로 대체하게 되었습니다. 멘토링 시간에 말씀해주셨던 부분을 참고해서 해결책을 찾아보도록 하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/enumeration/Language.java:
```diff
@@ -0,0 +1,22 @@
+    차후에 필요시 다른 언어 추가
+ */
+public enum Language {
+    ARABIC,
```
이런 언어리스트는 DB에 저장하는게 추후 관리에 용이할 것 같습니다~

msugo1 replied:
오늘 멘토링 시간에 조언해주신 것처럼, 어렵겠지만 한 번 적용해보도록 하겠습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/repository/LanguageRepository.java:
```diff
@@ -0,0 +1,28 @@
+
+@Repository
+@RequiredArgsConstructor
+public class LanguageRepository {
```
`@Mapper`와 `@Repository`를 나누셨는데 현재 상황에서 굳이 나눌 필요가 있을까요?

msugo1 replied:
안그래도 `UserService` 를 구현할 때부터 `Mapper 클래스`와 `Repository 클래스`가 역할이 아예 똑같아서 이게 필요한가 고민하고 있었는데요.

 Mapper는 DB 자체를 담당하고 Repository는 DAO의 전달 같은 DB와의 `communication` 을 담당한다고 생각을 하게 되었는데 다시 돌아보니 `Mapper Interface`가 중간에서 DAO 전달을 해주고 쿼리와 직접적으로 관련된 부분은 `*Mapper.xml` 파일이네요. 

`LanguageRepository` 그리고 `UserRepository` 까지도 `Mapper Interface`와 중복되는 기능을 갖고 있기 때문에  굳이 필요 없을 듯 합니다.

`UserRepository`  삭제는 이슈를 하나 더 파서 작업하는게 나을까요?

f-lab-dev commented on src/main/java/me/soo/helloworld/service/LanguageService.java:
```diff
@@ -0,0 +1,123 @@
+        - 구사가능 언어(CAN_SPEAK): 4개
+        - 모국어(NATIVE): 4개
+
+        예상 예외 시나리오
+        - 추가 가능한 최대 언어의 개수가 초과했을 경우
+        - 중복된 언어가 추가된 경우
+            1) 새롭게 요청 받은 언어 목록 내에서 중복 언어가 존재하는 경우
+            2) 기존 DB 에 저장된 전체 언어 목록과 비교했을 때 중복 요청이 들어온 경우
+        - 언어 레벨이 맞지 않는 경우
+            1) 추가할 언어의 status 는 NATIVE 인데, level 이 NATIVE 로 지정되어 있지 않은 경우
+            2) 추가할 언어의 status 는 NATIVE 가 아닌데, level 이 NATIVE 로 지정되어 있는 경우
```
이런 것들을 테스트로 작성해보면 좋을 것 같아요~ 테스트가 좋은 명세서가 되어줄 수 있을 것 같네요

f-lab-dev commented on src/main/java/me/soo/helloworld/service/LanguageService.java:
```diff
@@ -0,0 +1,123 @@
+        2. Status 가 Native 가 아닌 경우 (CAN_SPEAK or LEARNING)
+        - 어떤 언어도 레벨이 NATIVE 가 되어서는 안됩니다.
+     */
+    private void checkProperLevel(List<LanguageData> newLangData, LanguageStatus status) {
```
여기서는 유효성검사이므로 `check` 보다는 `validate`라고 해주는게 좋을 것 같습니다~

f-lab-dev commented on src/main/resources/mappers/LanguageMapper.xml:
```diff
@@ -0,0 +1,30 @@
+
+    <insert id="insertLanguages" parameterType="map">
+
+        INSERT INTO speak (userId, language, level, status)
```
bulk insert를 구현하셨는데 이거를 사용하신 이유가 무엇이실까요?

msugo1 replied:
처음에는 하나씩 추가하는 기능만 구현하려고 했는데요. 찾아보니 DB에 데이터를 삽입할 때 데이터를 삽입하는데 드는 비용보다 `DB 연결`,  `서버에 쿼리 보내기`,  `SQL 구문을 파싱`에 따른 비용이 훨씬 크더라구요.

그래서,

- 소규모 트래픽에서는 사실 문제가 없겠지만 저희는 대용량 트래픽을 목표로 하고 있기 때문에 다수의 사용자가 언어를 하나씩 계속 추가하는 상황이 와버리면 그 비용이 감당이 안될 것이라는 생각이 들었습니다.

- 또한, 한 사용자가 최대로 추가가능한 `언어 개수의 제한`을 걸어두었기 때문에, 최대 개수를 한 번에 추가하더라도 삽입에 따른 비용은 크지 않고, DB 연결은 한 번만 이루어지기 때문에 많은 비용을 절감할 수 있을 것이라 생각했습니다.
 
- 덤으로, 다수의 언어를 한 번에 추가하는 기능을 추가하면 사용자들이 일일이 언어를 하나 씩 번거롭게 추가할 필요가 없으므로 `편의성`도 확보할 수 있을 것이라고 판단했습니다.

Commit:
[#10] 피드백 반영

LanguageController (modified)

* addLanguage 메소드
- 리턴타입void로 변경(불필요한 리턴값 생략)
- `@RequestParam` 제거, wrapper 클래스를 사용해 `@RequestBody`가 모든
  요청 값을 매핑할 수 있도록 함

Model (modified)

* Language (Enum, deleted)
- 언어이름 관련 데이터를 DB에 저장해서 관리하기 위해 불필요한 클래스
  언어이름 상수를 담은 Enum 클래스는 삭제

* LanguageData (modified)
- 인스턴스 변수 변경 (Language name -> Integer id)
  : DB에 언어 이름 대신에 언어 id를 대신 보관
  (언어 이름에 대한 정보를 얻어야 하는 경우에는 DB에서 언어정보 테이블과 `JOIN` 해서 값을 얻도록 함)

* LanguageDataWrapper (new)
- `@RequestBody`로 들어온 요청을 한번에 매핑하기 위한 `Wrapper` 클래스
- 변수로 `LanguageData 목록`과 추가하려는 언어의 `Status`를 가짐

LanguageService (modified)

* checkProperLevel 메소드
- 메소드명 변경 (checkProperLevel -> ValidateLevel)
- 로컬 변수명 변경 (isProperLevel -> isLevelValid)

LanguageRepository (deleted)

* `LanguageMapper` 인터페이스와 완전히 동일한 역할을 하고 있던
  불필요한 클래스 삭제

LanguageMapper.xml (modified)

* `speak` 테이블의 language 칼럼의 자료형을 이름을 담을 CHAR에서 id를 담을 INT로 변경함에 따라 쿼리 내의 변수명 변경

테스트 관련

- 위의 수정사항을 반영하기 위해 테스트 수정
- 테스트를 위한 언어의 id 값을 담은 상수 클래스 생성
  : 내부의 상수 값들에는 접근제한자를 설정하지 않아 해당 패키지(언어
  테스트) 외에
  상수 값들이 사용되는 것을 방지

f-lab-dev commented on src/test/java/me/soo/helloworld/service/language/TestLangId.java:
```diff
@@ -0,0 +1,33 @@
+
+    final static int KOREAN = 1;
+
+    final static int English = 2;
```
여기만 소문자인 이유가 있으실까요?

msugo1 replied:
여러 개를 타이핑 하다보니 실수했네요. 좀 더 신경썼어야 했는데 주의하겠습니다. 수정해서 올렸습니다!

Commit:
[#10] 피드백 반영

* 상수값 오탈자 수정
- 소문자가 섞여 있던 상수 값 수정
